[iOS] Specify a default size for UISearchBar width if needed

rmarinho opened the pull request:
### Description of Change ###

As of latest versions of iOS11.3 seems that passing the max value to the UISearchBar for sizing breaks.
This only happens when we don't specify any Width request for the SearchBar and the SearchBar is requested to stretch itself. 
We could try to pass a enormous value but I figure this was a best call. 

At the same time we found out iOS10 also wasn't rendering , the issue that is SizeThatFits always is retiring a 0 Width , we override the call and set the width. 

### Issues Resolved ###

- fixes #3413 
- fixes #2139  

### API Changes ###

None

### Platforms Affected ###

- iOS

### Behavioral/Visual Changes ###

Should show the Searchbar inside a horizontal StackLayout and blue SearchBar Expanding

Before:

It crashed  and didn't renderer the blue SearchBar with CenterAndExpand

After:

IOS 10
![simulator screen shot - iphone 5 - 2018-08-29 at 12 38 38](https://user-images.githubusercontent.com/1235097/44786344-b88f4100-ab8b-11e8-982b-f49f0ff96098.png)
![simulator screen shot - iphone 5 - 2018-08-29 at 12 38 46](https://user-images.githubusercontent.com/1235097/44786347-b927d780-ab8b-11e8-8c52-5b5af68dcd63.png)

iOS11
![simulator screen shot - iphone x - 2018-08-29 at 12 20 30](https://user-images.githubusercontent.com/1235097/44786348-b927d780-ab8b-11e8-9742-7749febf409a.png)
![simulator screen shot - iphone x - 2018-08-29 at 12 20 44](https://user-images.githubusercontent.com/1235097/44786349-b927d780-ab8b-11e8-8d51-34d370952729.png)

### PR Checklist ###

- [ ] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [ ] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard

samhouts commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3413.cs:
```diff
@@ -0,0 +1,66 @@
+
+#if UITEST
+		[Test]
+		public void Issue3413Test ()
```
maybe add `ManualReview` test category

samhouts commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3413.cs:
```diff
@@ -0,0 +1,66 @@
+		{
+			RunningApp.WaitForElement (q => q.Marked ("srb_vertical"));
+			RunningApp.WaitForElement (q => q.Marked ("srb_horizontal"));
+			RunningApp.Screenshot ("Please verify we have 2 SearchBar's. One below the label, other side by side with the label");
```
SearchBars (no apostrophe)

rmarinho commented:
Hey @PureWeen  thanks for the test on older devices.

I had a look and I got it to work on iOS10, not sure if it worked without this on the pass or not.. but basically SizeToFit returns different values , for example is always return width of 0 even if you pass for example 300 on iOS10.. So the trick is set the with after and not call the base SizeThatFits of the renderer. 

At the same time i see that after rotating is not filling all space on iOS10 (on the right) I might investigate further that to see if I can fix that too. 

iOS 11 left, ios10 right 

<img width="1198" alt="screen shot 2018-08-23 at 12 32 33" src="https://user-images.githubusercontent.com/1235097/44523974-cba99900-a6d3-11e8-815e-3e0599b07e56.png">

<img width="1642" alt="screen shot 2018-08-23 at 12 52 30" src="https://user-images.githubusercontent.com/1235097/44523996-dcf2a580-a6d3-11e8-9dda-6b4dd18df86f.png">

rmarinho commented:
Found #2139 I bet is related

samhouts commented:
@rui 
> Found #2139 I bet is related
Does that mean this PR should fix that issue?

rmarinho commented:
Yes it does @samhouts I updated the screenshots and issue to show the "blue" searchbar that's the one using CenterAndExpand.

@PureWeen I tried another fix, and I think is better , I also updated screenshots. Please review. Thanks

The big issue is SizeThatFits is always broken on iOS10 for the with, passing a value, infinity or 0 will always return values that are not consistent with iOS11.

kingces95 commented:
Well, I tried these modification https://github.com/xamarin/Xamarin.Forms/pull/3686 but I'm seeing odd behavior all around. When I rotate the device the code in question is not always called tho the visual effect is different between my proposed changes and your code. So... I dunno what's going on. Anyway,

I find the trinay statement hard to grock. Tried to split it up.

The comment talks about a zero value so I expected the code to check for zero.

Tried to generalize the version check for all versions. Fewer version checks the better IMHO. 

But like I said, my code works when it's called but it's not always called on rotation for <11 devices.
